### PR TITLE
introduces `onClose` listener for RSocketClient and `connect` method

### LINKED ISSUE #1048: Support close in RSocketClient that notifies of the result
`RSocketClient` implements `Disposable` which provides a way to initiate a shutdown. However, there is no good way to hold for the result, and it is important in some cases to wait for that. For example, in components that expose start/stop lifecycle methods and could be started and stopped multiple times. Likewise in tests to ensure resources are released before continuing.

By comparison `RSocket` implements `Closeable` so technically possible to access the underlying source `RSocket` and call `close()` on that but that could cause a session to be started if not already connected. It would be better to align the two and expose a similar option at the level of `RSocketClient` as well.

### PR DESCRIPTION
**Pull request by OlegDokuka**
closes #1048 

Signed-off-by: Oleh Dokuka <odokuka@vmware.com>
Signed-off-by: Oleh Dokuka <oleh.dokuka@icloud.com>

### CODE REVIEW

**Review by rstoyanchev — COMMENTED** [review]
Looks good. Only a couple of small comments for the `connect` method, which could also be split out into a separate commit as it is orthogonal to the `onClose` I think.

File: rsocket-core/src/main/java/io/rsocket/core/RSocketClient.java [hunk 1]
```diff
@@ -74,7 +75,15 @@
  * @since 1.1
  * @see io.rsocket.loadbalance.LoadbalanceRSocketClient
  */
-public interface RSocketClient extends Disposable {
+public interface RSocketClient extends Closeable {
+
+  default boolean connect() {
+    return false;
+  }
```
[the remark this answers is not in the document]
Comment (OlegDokuka) [reply rsocket-core/src/main/java/io/rsocket/core/RSocketClient.java:88]: fixed

**Comment by OlegDokuka** [comment]
@rstoyanchev added 2 separate commits as it was suggested

File: rsocket-core/src/main/java/io/rsocket/core/RSocketClient.java [hunk 2]
```diff
@@ -74,7 +75,21 @@
  * @since 1.1
  * @see io.rsocket.loadbalance.LoadbalanceRSocketClient
  */
-public interface RSocketClient extends Disposable {
+public interface RSocketClient extends Closeable {
+
+  /**
+   * Tries to connect to remote rsocket endpoint if not yet connected. This method is a
+   * shortcut to {@code RSocketClient#source().subscribe()}
```
Comment (rstoyanchev) [inline rsocket-core/src/main/java/io/rsocket/core/RSocketClient.java:82]:
```suggestion
   * Connect to the remote rsocket endpoint, if not yet connected. This method is a
   * shortcut for {@code RSocketClient#source().subscribe()}.
```

File: rsocket-core/src/main/java/io/rsocket/core/RSocketClient.java [hunk 3]
```diff
@@ -74,7 +75,21 @@
...
+   *
+   * @return {@code true} if successfully activated connection process
```
Comment (rstoyanchev) [inline rsocket-core/src/main/java/io/rsocket/core/RSocketClient.java:84]:
```suggestion
   * @return {@code true} if an attempt to connect was triggered, or {@code false}
   * if already connected or if the client is terminated.
```

File: rsocket-core/src/main/java/io/rsocket/core/RSocketClientAdapter.java [hunk 4]
```diff
@@ -41,11 +41,22 @@ public RSocket rsocket() {
     return rsocket;
   }
 
+  @Override
+  public boolean connect() {
+    throw new UnsupportedOperationException(
+        "Serverside RSocketClientAdapter can not proactively open a connection");
```
Comment (rstoyanchev) [inline rsocket-core/src/main/java/io/rsocket/core/RSocketClientAdapter.java:47]:
```suggestion
        "Connect does not apply to a server side RSocket");
```

**Review by rstoyanchev — APPROVED** [review]

**Pull request merged**